x/exchange/keeper/market.go

```diff
@@ -0,0 +1,1303 @@
+	var errs []error
+
+	for _, addrStr := range msg.RevokeAll {
+		addr := sdk.MustAccAddressFromBech32(addrStr)
+		perms := getUserPermissions(store, marketID, addr)
+		if len(perms) > 0 {
+			if len(errs) == 0 {
+				revokeUserPermissions(store, marketID, addr)
+			}
+		} else {
+			errs = append(errs, fmt.Errorf("account %s does not have any permissions for market %d", addrStr, marketID))
+		}
+	}
+
+	for _, ag := range msg.ToRevoke {
+		addr := sdk.MustAccAddressFromBech32(ag.Address)
+		for _, perm := range ag.Permissions {
+			if !storeHasPermission(store, marketID, addr, perm) {
+				errs = append(errs, fmt.Errorf("account %s does not have %s for market %d", ag.Address, perm.String(), marketID))
+			}
+		}
+		if len(errs) == 0 {
+			revokePermissions(store, marketID, addr, ag.Permissions)
+		}
+	}
+
+	for _, ag := range msg.ToGrant {
+		addr := sdk.MustAccAddressFromBech32(ag.Address)
+		for _, perm := range ag.Permissions {
+			if storeHasPermission(store, marketID, addr, perm) {
+				errs = append(errs, fmt.Errorf("account %s already has %s for market %d", ag.Address, perm.String(), marketID))
+			}
+		}
+		if len(errs) == 0 {
+			grantPermissions(store, marketID, addr, ag.Permissions)
+		}
+	}
+
+	if len(errs) > 0 {
+		return errors.Join(errs...)
+	}
+
+	k.emitEvent(ctx, exchange.NewEventMarketPermissionsUpdated(marketID, admin))
+	return nil
+}
+
+// reqAttrKeyMaker is a function that returns a key for required attributes.
+type reqAttrKeyMaker func(marketID uint32) []byte
+
+// getReqAttrs gets the required attributes for a market using the provided key maker.
+func getReqAttrs(store sdk.KVStore, marketID uint32, maker reqAttrKeyMaker) []string {
+	key := maker(marketID)
+	value := store.Get(key)
+	return ParseReqAttrStoreValue(value)
+}
+
+// setReqAttrs sets the required attributes for a market using the provided key maker.
+func setReqAttrs(store sdk.KVStore, marketID uint32, reqAttrs []string, maker reqAttrKeyMaker) {
+	key := maker(marketID)
+	if len(reqAttrs) == 0 {
+		store.Delete(key)
+	} else {
+		value := []byte(strings.Join(reqAttrs, string(RecordSeparator)))
+		store.Set(key, value)
+	}
+}
+
+// updateReqAttrs updates the required attributes in the store that use the provided key maker by removing then adding
+// the provided attributes to the existing entries.
+func updateReqAttrs(store sdk.KVStore, marketID uint32, toRemove, toAdd []string, field string, maker reqAttrKeyMaker) error {
+	var errs []error
+	curAttrs := getReqAttrs(store, marketID, maker)
+
+	for _, attr := range toRemove {
+		if !exchange.ContainsString(curAttrs, attr) {
+			errs = append(errs, fmt.Errorf("cannot remove %s required attribute %q: attribute not currently required", field, attr))
+		}
+	}
+
+	var updatedAttrs []string
+	for _, attr := range curAttrs {
+		if !exchange.ContainsString(toRemove, attr) {
+			updatedAttrs = append(updatedAttrs, attr)
+		}
+	}
+
+	for _, attr := range toAdd {
+		if !exchange.ContainsString(curAttrs, attr) {
```

Comment:
If you flip the logic here then you can prevent the happy path from being indented.

```
if exchange.ContainsString(curAttrs, attr) {
  errs = append(errs, fmt.Errorf("cannot add %s required attribute %q: attribute already required", field, attr))
  continue
}
updatedAttrs = append(updatedAttrs, attr)
```

Reply:
I much prefer an else to a continue.